src/lib.rs

```diff
@@ -29,3 +35,118 @@ mod violation;
 pub mod external {
     pub use ion_rs;
 }
+
+/// Provide an Ion schema element which includes all OwnedElements and a document type
+///
+/// ## Example:
+/// In general `TypeRef` `validate()` takes in IonSchemaElement as the value to be validated.
+/// In order to create an `IonSchemaElement`:
+///
+/// ```
+/// use ion_rs::value::owned::OwnedElement;
+/// use ion_schema::IonSchemaElement;
+///
+/// // create an IonSchemaElement from an OwnedElement
+/// let owned_element: OwnedElement = 4.into();
+/// let ion_schema_element: IonSchemaElement = (&owned_element).into();
+///
+/// // create an IonSchemaElement for document type based on vector of owned elements
+/// let document: IonSchemaElement = IonSchemaElement::Document(vec![owned_element]);
+/// ```
+#[derive(Debug, Clone)]
+pub enum IonSchemaElement {
+    SingleElement(OwnedElement),
+    Document(Vec<OwnedElement>),
+}
+
+impl IonSchemaElement {
+    pub fn as_element(&self) -> Option<&OwnedElement> {
+        match self {
+            IonSchemaElement::SingleElement(element) => Some(element),
+            IonSchemaElement::Document(_) => None,
+        }
+    }
+
+    fn expect_element_of_type(
+        &self,
+        constraint_name: &str,
+        types: &[IonType],
+    ) -> Result<&OwnedElement, Violation> {
+        match self {
+            IonSchemaElement::SingleElement(element) => {
+                if !types.contains(&element.ion_type()) || element.is_null() {
+                    // If it's an OwnedElement but the type isn't one of `types`,
+                    // return a Violation with the constraint name.
+                    return Err(Violation::new(
+                        constraint_name,
+                        ViolationCode::TypeMismatched,
+                        &format!("expected {:?} but found {}", types, element.ion_type()),
+                    ));
+                }
+                // If it's an OwnedElement of an expected type, return a ref to it.
+                Ok(element)
+            }
+            IonSchemaElement::Document(_) => {
+                // If it's a Document, return a Violation with the constraint name
+                Err(Violation::new(
+                    constraint_name,
+                    ViolationCode::TypeMismatched,
+                    &format!("expected {:?} but found document", types),
+                ))
+            }
+        }
+    }
+}
+
+impl Display for IonSchemaElement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        match self {
+            IonSchemaElement::SingleElement(element) => {
+                write!(f, "{}", element)
+            }
+            IonSchemaElement::Document(document) => {
+                write!(f, "document::( ")?;
```

Comment:
The `document::(...)` syntax could be confusing. We could do something like:
```
/* Ion document */ ... /* end */
```
(Yuck.) Definitely open to better alternatives. This doesn't need to block the PR since it's just for error reporting.

Reply:
Yes, I wad confused about this as this one is in particular an `SExpression` with `document` annotation. I can do this comment thing as I think that wouldn't conflict with any definition.